tests/functional_tests/ffront_tests/test_type_deduction.py

```diff
@@ -479,3 +480,69 @@ def badtype_broadcast(a: Field[[ADim], float64]):
         match=r"Expected all broadcast dimensions to be of type Dimension.",
     ):
         _ = FieldOperatorParser.apply_to_function(badtype_broadcast)
+
+
+def test_where_dim():
+    ADim = Dimension("ADim")
+    BDim = Dimension("BDim")
+
+    def simple_where(a: Field[[ADim], bool], b: Field[[ADim, BDim], float64]):
+        return where(a, b, 9.0)
+
+    parsed = FieldOperatorParser.apply_to_function(simple_where)
+
+    assert parsed.body[0].value.type == ct.FieldType(
+        dims=[ADim, BDim], dtype=ct.ScalarType(kind=ct.ScalarKind.FLOAT64)
+    )
+
+
+def test_where_broadcast_dim():
+    ADim = Dimension("ADim")
+
+    def simple_where(a: Field[[ADim], bool]):
+        return where(a, 5.0, 9.0)
+
+    parsed = FieldOperatorParser.apply_to_function(simple_where)
+
+    assert parsed.body[0].value.type == ct.FieldType(
+        dims=[ADim], dtype=ct.ScalarType(kind=ct.ScalarKind.FLOAT64)
+    )
+
+
+def test_where_tuple_dim():
+    ADim = Dimension("ADim")
+
+    def tuple_where(a: Field[[ADim], bool], b: Field[[ADim], float64]):
+        return where(a, ((5.0, 9.0), (b, 6.0)), ((8.0, b), (5.0, 9.0)))
+
+    parsed = FieldOperatorParser.apply_to_function(tuple_where)
+
+    assert parsed.body[0].value.type == ct.TupleType(
+        types=[
+            ct.TupleType(
+                types=[
+                    ct.FieldType(dims=[ADim], dtype=ct.ScalarType(kind=ct.ScalarKind.FLOAT64)),
+                    ct.FieldType(dims=[ADim], dtype=ct.ScalarType(kind=ct.ScalarKind.FLOAT64)),
+                ]
+            ),
+            ct.TupleType(
+                types=[
+                    ct.FieldType(dims=[ADim], dtype=ct.ScalarType(kind=ct.ScalarKind.FLOAT64)),
+                    ct.FieldType(dims=[ADim], dtype=ct.ScalarType(kind=ct.ScalarKind.FLOAT64)),
+                ]
+            ),
+        ]
+    )
+
```

Comment:
The following **crucial** test is missing:

```python
ADim = Dimension("ADim")
BDim = Dimension("BDim")

def tuple_where(a: Field[[ADim], bool], b: Field[[ADim], float64], c: Field[[ADim, BDim], float64]):
    return where(a, ((c, 9.0), (b, 6.0)), ((8.0, b), (5.0, 9.0)))
```

I am not 100 % sure what the return type should be but I would expect something like

```python
(
    (Field[[ADim, BDim], ...], Field[[ADim], ...]),
    (Field[[ADim], ...], Field[[ADim], ...]),
)
```

In any case looking at the code I don't think the question this test asks has been addressed: "What happens if I mix dimensions in a nested tuple (in a way that the two branches are compatible)?".

Reply:
As the implementation is atm, the return dimensions should be [ADim, BDim] for all tuples items.

Reply:
Yes, I think we should discuss this case together.